src/planner/binder/statement/bind_create_table.cpp

```diff
@@ -144,12 +175,132 @@ static void BindConstraints(Binder &binder, BoundCreateTableInfo &info) {
+		D_ASSERT(success);
+		if (table->types[index] == LogicalType::ANY) {
+			unresolved = true;
+			return;
+		}
+	}
+	ParsedExpressionIterator::EnumerateChildren(
+	    expr, [&](const ParsedExpression &child) { TypeIsUnresolved((ParsedExpression &)child, table, unresolved); });
+}
+
+static bool HasUnresolvedDependency(column_t index, const CreateTableInfo &info, unordered_set<column_t> &resolved) {
+	if (!info.column_dependency_manager.HasDependencies(index)) {
+		return false;
+	}
+	auto &dependencies = info.column_dependency_manager.GetDependencies(index);
+	for (auto &dep : dependencies) {
+		if (!info.column_dependency_manager.HasDependencies(dep)) {
+			continue;
+		}
+		// A generated column and not resolved yet
+		if (!resolved.count(dep)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+struct GeneratedColumnBindData {
+	column_t index;
+	bool allow_unresolved;
+};
+
+void Binder::BindGeneratedColumns(vector<ColumnDefinition> &columns, const CreateTableInfo &info) {
+	vector<string> names;
+	vector<LogicalType> types;
+	queue<GeneratedColumnBindData> to_resolve;
+
+	D_ASSERT(info.type == CatalogType::TABLE_ENTRY);
+	for (idx_t i = 0; i < info.columns.size(); i++) {
+		auto &col = info.columns[i];
+		names.push_back(col.name);
+		types.push_back(col.type);
+		if (!col.Generated()) {
+			continue;
+		}
+		to_resolve.push({i, true});
+	}
+	// Nothing to do
+	if (to_resolve.empty()) {
+		return;
+	}
+	auto table_index = GenerateTableIndex();
+
+	// Create a new binder because we dont need (or want) these bindings in this scope
+	auto binder = Binder::CreateBinder(context);
+	binder->bind_context.AddGenericBinding(table_index, info.table, names, types);
+	string ignore;
+	auto table_binding = binder->bind_context.GetBinding(info.table, ignore);
+	D_ASSERT(table_binding && ignore.empty());
+
+	unordered_set<column_t> resolved;
+
+	// Continue until all generated columns are resolved
+	for (; !to_resolve.empty();) {
+		auto resolve_data = to_resolve.front();
+		to_resolve.pop();
+		column_t i = resolve_data.index;
+		// Already resolved
+		bool allow_unresolved = resolve_data.allow_unresolved;
+		auto &col = columns[i];
+		if (resolved.count(i)) {
+			continue;
+		}
+
+		auto expr_binder = ExpressionBinder(*binder, context);
+		auto expression = col.GeneratedExpression().Copy();
+		if (HasUnresolvedDependency(i, info, resolved)) {
+			if (allow_unresolved) {
+				// Add it to the back of the queue
+				to_resolve.push({i, false});
+				continue;
+			}
+			//! When it gets resolved next time and it still has unresolved dependencies, we throw an error
+			throw InvalidInputException("Circular dependency encountered when resolving generated column expression");
+		}
+		// expr_binder.target_type = col.type;
+		auto bound_expression = expr_binder.Bind(expression);
+		if (!bound_expression) {
+			throw BinderException("Could not resolve the expression of generated column \"%s\"", col.name);
+		}
+		if (bound_expression->HasSubquery()) {
```

Comment:
The `CheckBinder` does this as well

Reply:
I've been able to get rid of many of the logic I was duplicating, but the CheckBinder mostly calls ExpressionBinder::BindExpression and the changing of the column type in the case of an optional type is something unique to generated expression binding